game-core/src/main/java/games/strategy/triplea/image/UnitImageFactory.java

```diff
@@ -120,26 +106,29 @@ public Image getImage(final UnitCategory unit) {
       final UnitType type, final GamePlayer player, final boolean damaged, final boolean disabled) {
     final String baseName = getBaseImageName(type, player, damaged, disabled);
     final String fullName = baseName + player.getName();
-    if (images.containsKey(fullName)) {
-      return Optional.of(images.get(fullName));
-    }
-    final Optional<Image> image = getTransformedImage(baseName, player, type);
-    if (image.isEmpty()) {
-      return Optional.empty();
-    }
-    final Image baseImage = image.get();
 
-    // We want to scale units according to the given scale factor.
-    // We use smooth scaling since the images are cached to allow to take our time in doing the
-    // scaling.
-    // Image observer is null, since the image should have been guaranteed to be loaded.
-    final int width = (int) (baseImage.getWidth(null) * scaleFactor);
-    final int height = (int) (baseImage.getHeight(null) * scaleFactor);
-    final Image scaledImage = baseImage.getScaledInstance(width, height, Image.SCALE_SMOOTH);
-    // Ensure the scaling is completed.
-    Util.ensureImageLoaded(scaledImage);
-    images.put(fullName, scaledImage);
-    return Optional.of(scaledImage);
+    return Optional.ofNullable(images.get(fullName))
```

Comment:
Actually I'd go for `Optional.ofNullable(images.computeIfAbsent(fullName, k -> {/* The lambda below*/}))`.

However this change would no longer attempt to recompute an image if `getTransformedImage(baseName, player, type)` returned an empty optional, so maybe we need something like this instead:
```java
Optional.ofNullable(images.compute(fullName, (key, image) -> Optional.ofNullable(image).or(
/* lambda below*/
).orElse(null)));
```

Reply:
Extracting the computation function to another method helps (arguably the thing to do to avoid too much inline code), and `computeIfAbsent` does avoid the optional `or`.

Updated: 1fdf91d4b

Reply:
Interesting landed quickly into concurrent modification errors, reverted the update: cc3e02409